### PR TITLE
Migrate to JupyterLab 4 / Lumino 2

### PR DESCRIPTION
**Pull request by vthemelis**
Also fixes https://github.com/bloomberg/ipydatagrid/issues/431

### CODE REVIEW

**Comment by martinRenou** [comment]
Thank you for opening a PR! Unfortunately we'll be stuck waiting for ipywidgets to build against lab 4 before we can update https://github.com/jupyter-widgets/ipywidgets/pull/3752

**Comment by martinRenou** [comment]
I think we can now resume this work

File: js/datagrid.ts [hunk 1]
```diff
@@ -189,10 +189,11 @@ export class DataGridModel extends DOMWidgetModel {
...
         this.synchingWithKernel = true;
 
-        const selectionIter = sender.selections().iter();
+        const selectionIter = sender.selections();
         const selections: any[] = [];
-        let selection = null;
-        while ((selection = selectionIter.next())) {
+        let selectionNode = null;
+        while ((selectionNode = selectionIter.next())) {
+          const selection = selectionNode.value;
```
Comment (vthemelis) [inline js/datagrid.ts:196]: This fixes https://github.com/bloomberg/ipydatagrid/issues/431

**Comment by vthemelis** [comment]
@martinRenou, could you help me fix the Galata tests? I'm not super familiar with those.

**Comment by martinRenou** [comment]
Sure I'll try to get your PR green. Thanks!

File: js/datagrid.ts (hunk 1 above)
Comment (vthemelis) [reply js/datagrid.ts:196]: But is backwards incompatible with older versions of lumino which I personally think is fine but probably means we should bump the minor version.

**Comment by vthemelis** [comment]
@martinRenou, I think that this is now read for review. You will also have to update the UI test artefacts (I don't think I can do that) but the tests run just fine.

**Comment by martinRenou** [comment]
Thanks! I'll have a look very soonish, it's on my todo list

**Comment by martinRenou** [comment]
Looking into it this afternoon!

**Comment by martinRenou** [comment]
Turns out the CI testing against ipywidgets 7 catches a real issue. ipydatagrid does not render when using ipywidgets 7 with this PR. I'm investigating.

**Comment by martinRenou** [comment]
After more investigation, it's not an ipydatagrid issue but an ipywidgets issue. ipywidgets 7.x does not work with JupyterLab 4.

We will want to report this to the ipywidgets repo. I don't know if it's possible to make a patch version of ipywidgets 7.x fixing this.

I will update our visual regression tests to use JupyterLab 3 in ui-tests-ipw7 and JupyterLab 4 in ui-tests-ipw8.

File: ipydatagrid/_version.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
 # Copyright (c) Bloomberg.
 # Distributed under the terms of the Modified BSD License.
 
-__version__ = "1.1.16"
```
Comment (martinRenou) [inline ipydatagrid/_version.py:4]: We should bump this separately

File: js/datagrid.ts (hunk 1 above)
Comment (martinRenou) [reply js/datagrid.ts:196]: Let's try making something that works with both. I will look into it.

File: ipydatagrid/_version.py (hunk 2 above)
Comment (martinRenou) [reply ipydatagrid/_version.py]: Done

File: js/datagrid.ts (hunk 1 above)
Comment (martinRenou) [reply js/datagrid.ts]: Done!

File: ipydatagrid/_version.py (hunk 2 above)
Comment (vthemelis) [reply ipydatagrid/_version.py]: The package.json still has the new version I think

**Review by martinRenou — APPROVED** [review]
Thanks for your work!

**Pull request merged**